no-unused-vars argsIgnorePattern affecting "args": "after-used" since 3.6.0

**Tell us about your environment**

* **ESLint Version:** v3.6.0
* **Node Version:** v6.6.0
* **npm Version:** 3.10.7

**What parser (default, Babel-ESLint, etc.) are you using?** deafult

**Please show your full configuration:** https://github.com/moodle/moodle/blob/master/.eslintrc

**What did you do? Please include the actual source code causing the issue.**

When attempting to upgrade to eslint 3.6.0, we got some new errors reported by no-unused-vars which should have been ignored by `args: 'after-used'`. It turned out that argsIgnorePattern is now affecting the result of things which should be ignored by `"args": "after-used"`

I attempted to isolate the code down to the smallest test case:

Problem file:
```js
/* eslint 'no-unused-vars': ['error', {"args": "after-used", "argsIgnorePattern": "^baz$"}] */

(function(foo, bar, baz) {
        return baz;
})();
```

Output:
```sh
$ eslint problem.js 

/Users/danp/www/eslintest/problem.js
  3:16  error  'bar' is defined but never used  no-unused-vars

✖ 1 problem (1 error, 0 warnings)
```

Ok file:
```js
/* eslint 'no-unused-vars': ['error', {"args": "after-used"}] */

(function(foo, bar, baz) {
        return baz;
})();
```

--- a/lib/rules/no-unused-vars.js
+++ b/lib/rules/no-unused-vars.js
@@ -428,12 +428,12 @@ module.exports = {
                 return true;
             }
 
-            // if all parameters preceded by this variable are ignored, this is the last.
+            // if all parameters preceded by this variable are ignored and unused, this is the last.
             if (config.argsIgnorePattern) {
                 const params = context.getDeclaredVariables(def.node);
                 const posteriorParams = params.slice(params.indexOf(variable) + 1);
 
-                if (posteriorParams.every(v => config.argsIgnorePattern.test(v.name))) {
+                if (posteriorParams.every(v => v.references.length === 0 && config.argsIgnorePattern.test(v.name))) {
                     return true;
                 }
             }

--- a/tests/lib/rules/no-unused-vars.js
+++ b/tests/lib/rules/no-unused-vars.js
@@ -225,6 +225,17 @@ ruleTester.run("no-unused-vars", rule, {
             options: [{argsIgnorePattern: "d"}],
             parserOptions: {ecmaVersion: 6}
         },
+
+        // https://github.com/eslint/eslint/issues/7250
+        {
+            code: "(function(a, b, c) { c })",
+            options: [{argsIgnorePattern: "c"}],
+        },
+        {
+            code: "(function(a, b, {c, d}) { c })",
+            options: [{argsIgnorePattern: "[cd]"}],
+            parserOptions: {ecmaVersion: 6},
+        },
     ],
     invalid: [
         { code: "function foox() { return foox(); }", errors: [{ message: "'foox' is defined but never used.", type: "Identifier"}] },

